cpp/src/arrow/compute/kernels/vector_replace.cc

```diff
@@ -442,23 +442,409 @@ struct ReplaceWithMaskFunctor {
+    FillNullInDirectionImpl<Type>(array, reversed_bitmap, output, direction,
+                                  last_valid_value_chunk, last_valid_value_offset);
+    return Status::OK();
+  }
+};
+
+template <typename Type>
+struct FillNullExecutor<Type, enable_if_base_binary<Type>> {
+  using offset_type = typename Type::offset_type;
+  using BuilderType = typename TypeTraits<Type>::BuilderType;
+
+  static Status ExecFillNull(KernelContext* ctx, const ArrayData& array,
+                             const uint8_t* reversed_bitmap, ArrayData* output,
+                             int8_t direction, ArrayData last_valid_value_chunk,
+                             int64_t* last_valid_value_offset) {
+    BuilderType builder(array.type, ctx->memory_pool());
+    RETURN_NOT_OK(builder.Reserve(array.length));
+    RETURN_NOT_OK(builder.ReserveData(array.buffers[2]->size()));
+    int64_t array_value_index = direction == 1 ? 0 : array.length - 1;
+    const uint8_t* data = array.buffers[2]->data();
+    const uint8_t* data_prev = last_valid_value_chunk.buffers[2]->data();
+    const offset_type* offsets = array.GetValues<offset_type>(1);
+    const offset_type* offsets_prev = last_valid_value_chunk.GetValues<offset_type>(1);
+
+    bool has_fill_value_last_chunk = *last_valid_value_offset != -1;
+    bool has_fill_value_current_chunk = false;
+    std::vector<std::tuple<bool, uint64_t, uint64_t>> offsets_reversed;
+    RETURN_NOT_OK(VisitNullBitmapInline<>(
+        reversed_bitmap, array.offset, array.length, array.GetNullCount(),
+        [&]() {
+          const offset_type offset0 = offsets[array_value_index];
+          const offset_type offset1 = offsets[array_value_index + 1];
+          offsets_reversed.push_back(
+              std::make_tuple(/*current_chunk=*/1, offset0, offset1 - offset0));
+          *last_valid_value_offset = array_value_index;
+          has_fill_value_current_chunk = true;
+          has_fill_value_last_chunk = false;
+          array_value_index += direction;
+          return Status::OK();
+        },
+        [&]() {
+          if (has_fill_value_current_chunk || has_fill_value_last_chunk) {
+            if (!has_fill_value_last_chunk) {
+              const offset_type offset0 = offsets[*last_valid_value_offset];
+              const offset_type offset1 = offsets[*last_valid_value_offset + 1];
+              offsets_reversed.push_back(
+                  std::make_tuple(/*current_chunk=*/1, offset0, offset1 - offset0));
+            } else {
+              const offset_type offset0 = offsets_prev[*last_valid_value_offset];
+              const offset_type offset1 = offsets_prev[*last_valid_value_offset + 1];
+              offsets_reversed.push_back(
+                  std::make_tuple(/*current_chunk=*/0, offset0, offset1 - offset0));
+            }
+          } else {
+            offsets_reversed.push_back(std::make_tuple(/*current_chunk=*/0, -1U, -1U));
```

Comment:
If we're going to use negative numbers, then just make the offsets `int64_t`. Arrow generally uses signed offsets/lengths anyways.